driver/config/config.cpp

```diff
@@ -97,3 +99,270 @@ void writeDSNinfo(const ConnInfo * ci) {
+    std::list<std::basic_string<CharTypeLPCTSTR>> keys;
+
+    fromUTF8(dsn_utf8, dsn);
+    fromUTF8(ODBC_INI, config_file);
+
+    // Read all key names under the specified DSN.
+    {
+        std::basic_string<CharTypeLPCTSTR> keys_all(MAX_DSN_KEY_LEN * 1000, CharTypeLPCTSTR{});
+
+        const auto read = SQLGetPrivateProfileString(
+            dsn.c_str(),
+            nullptr,
+            default_value.c_str(),
+            const_cast<CharTypeLPCTSTR *>(keys_all.data()),
+            keys_all.size(),
+            config_file.c_str()
+        );
+
+        if (read < 0)
+            throw std::runtime_error("SQLGetPrivateProfileString failed to extract the key list of DSN " + dsn_utf8);
+
+        if (read >= keys_all.size())
+            throw std::runtime_error("SQLGetPrivateProfileString failed to extract the entire key list of DSN " + dsn_utf8);
+
+        keys_all.resize(read);
+
+        std::basic_string<CharTypeLPCTSTR> key;
+        for (auto ch : keys_all) {
+            if (ch == CharTypeLPCTSTR{}) {
+                if (!key.empty()) {
+                    keys.push_back(key);
+                    key.clear();
+                }
+            }
+            else {
+                key += ch;
+            }
+        }
+
+        if (!key.empty())
+            keys.push_back(key);
+    }
+
+// BEGIN WORKAROUND: some versions of UnixODBC instead of returning all keys return only the first one,
+// so here we make sure that all known keys will aslo be tried.
+    for (const auto & known_key :
+        {
+            INI_FILEDSN,
+            INI_SAVEFILE,
+            INI_DSN,
+            INI_DRIVER,
+            INI_DESC,
+            INI_URL,
+            INI_PROTO,
+            INI_SERVER,
+            INI_HOST,
+            INI_PORT,
+            INI_PATH,
+            INI_UID,
+            INI_USERNAME,
+            INI_PWD,
+            INI_PASSWORD,
+            INI_TIMEOUT,
+            INI_SSLMODE,
+            INI_DATABASE,
+            INI_READONLY,
+            INI_STRINGMAXLENGTH,
+            INI_PRIVATEKEYFILE,
+            INI_CERTIFICATEFILE,
+            INI_CALOCATION,
+            INI_DRIVERLOG,
+            INI_DRIVERLOGFILE
+        }
+    ) {
+        if (
+            std::find_if(
+                keys.begin(),
+                keys.end(),
+                [&] (const auto & key) { return (Poco::UTF8::icompare(known_key, toUTF8(key)) == 0); }
+            ) == keys.end()
+        ) {
+            keys.push_back(fromUTF8<CharTypeLPCTSTR>(known_key));
+        }
+    }
+// END WORKAROUND
+
+    // Read values of each key.
+    for (const auto key : keys) {
+        const auto key_utf8 = toUTF8(key);
+
+        if (fields.find(key_utf8) != fields.end()) {
+            LOG("Repeating key " << key_utf8 << " in DSN " << dsn_utf8 << ", ignoring, the first met value will be used");
+            continue;
+        }
+
+        std::basic_string<CharTypeLPCTSTR> value(MAX_DSN_VALUE_LEN, CharTypeLPCTSTR{});
```

Comment:
IMO, it would be much, much better (performance-wise) to have a plain c-style array as buffer, declared outside of for loop here.
```
CharTypeLPCTSTR value_buffer[MAX_DSN_VALUE_LEN] = {CharTypeLPCTSTR{}};
for (const auto key : keys) {
// ...
}

Reply:
I don't think that's a good idea, because 1) this loop is not performance critical at all, it is usually a loop of 5-10 iterations, that is entered only 1 time per connection (which usually means 1 time per client app start), 2) having C-style array would mean maintaining a separate length variable, and so on, which bring us to pretty much what `std::basic_string<...>` does. Some older code in other places used that approach, and it was a pain modifying it, during adding/fixing Unicode support. This proved to be more readable and maintainable.